Server-side schema ID validation (23.2 beta feature) (#1587)
redpanda-data/documentation: versioned_docs/version-23.2/manage/schema-id-validation.mdx

---
title: Server-side Schema ID Validation
linkRoot: ../../
---

<head>
    <meta name="title" content="Server-side Schema ID Validation | Redpanda Docs"/>
    <meta name="description" content="Learn about server-side schema ID validation for clients using Confluent's serdes format that produce to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy and registered with the Schema Registry."/>
    <link rel="canonical" href="https://docs.redpanda.com/manage/schema-registry/" />
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import EnterpriseLicenseNote from '@site/docs/shared/_enterprise-license.mdx'

> Learn about server-side schema ID validation for clients using Confluent's serdes format that produce to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy and registered with the Schema Registry.

Reviewer: ```suggestion
Learn about server-side schema ID validation for clients using Confluent's serdes format that produces to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy and registered with the Schema Registry.
```

Author: Changing to SerDes

:::info
<EnterpriseLicenseNote frontmatter={frontMatter}/>
:::

## About schema ID validation

> Records produced to a topic may use a serializer/deserializer client library, such as Confluent's serdes library, to encode their keys and values according to a schema in [Schema Registry](../schema-registry).

Reviewer: Serdes or SerDes

Schema ID validation enables brokers (servers) to detect and drop records that were produced with an incorrectly configured subject name strategy, that don't conform to the serdes wire format, or encode an incorrect schema ID. With schema ID validation, records associated with unregistered schemas are detected and dropped earlier, by a broker rather than a consumer. 

:::info
Schema ID validation doesn't verify that a record's payload is correctly encoded according to the associated schema. Schema ID validation only checks that the schema ID encoded in the record is registered in the Schema Registry.
:::

## Configure schema ID validation

To use schema ID validation:
- [Enable the feature in Redpanda](#enable-schema-id-validation)
- [Customize the subject name strategy per topic on the client](#set-subject-name-strategy-per-topic)

### Enable schema ID validation

To enable schema ID validation in Redpanda, set it with the `/v1/features/schema_id_validation` Admin API endpoint:

```bash
curl -sv -X PUT <broker-address>:9644/v1/features/schema_id_validation -d '{"state":"active"}'
```

### Set subject name strategy per topic

> The subject name strategies supported by the Redpanda:

Reviewer: ```suggestion
The subject name strategies supported by Redpanda:
```

---
title: Server-side Schema ID Validation
linkRoot: ../../
---

<head>
    <meta name="title" content="Server-side Schema ID Validation | Redpanda Docs"/>
    <meta name="description" content="Learn about server-side schema ID validation between clients using Confluent's serdes library and Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that don't match a schema in Schema Registry."/>
    <link rel="canonical" href="https://docs.redpanda.com/manage/schema-registry/" />
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import EnterpriseLicenseNote from '@site/docs/shared/_enterprise-license.mdx'

> Learn about server-side schema ID validation between clients using Confluent's serdes library and Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that don't match a schema in Schema Registry.

Reviewer 2: This is probably a bit more accurate. I'll let you wordsmith it and update the `<meta name="description"`
```suggestion
Learn about server-side schema ID validation for clients using Confluent's serdes format that produce to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy, and registered with the Schema Registry.
```

Records produced to a topic may use a serializer/deserializer (serde) client library, such as Confluent's serdes library, to encode their keys and values according to a schema in [Schema Registry](../schema-registry).

When a client produces a record, the schema ID for the topic is encoded in the record's payload header. A client may misconfigure and produce a record whose schema ID doesn't match any schema in Schema Registry. A produced record for an unregistered schema shouldn't be stored by brokers or fetched by consumers. Yet, it may not be detected or dropped until after it's been fetched and a consumer deserializes its mismatched schema ID.

**Schema ID validation enables brokers (servers) to detect and drop records whose encoded schema IDs don't match registered schemas**. With schema ID validation, records associated with unregistered schemas are detected and dropped earlier, by a broker rather than a consumer. 

:::info
Schema ID validation doesn't verify that a record's payload is correctly encoded according to the associated schema. Schema ID validation only checks that the schema ID encoded in the record is registered in the Schema Registry.

> :::

Reviewer 2: A point of clarification for these paragraphs:

The schema ID is encoded into the header of a record field, where a record field is the `key` or `value` of that record. The `topic` and `schema record` are combined according to the `Subject Name Strategy` to form a `Subject`. The Schema ID is associated with a `Subject` and `Version`. 

So the validation is a lot more strict than just "is there a schema registered with that ID". The validation is designed to detect producers that have an incorrectly configured `Subject Name Strategy`, or outright just don't conform to the serdes `Wire Format`, or otherhow manage to encode an incorrect schema ID.

| Subject Name Strategy | Subject Name Source | Subject Name Format (Key) | Subject Name Format (Value) |
|- | - | - | - |
| `io.confluent.kafka.serializers.subject.TopicNameStrategy` | Topic name | `<topic-name>-key` | `<topic-name>-value` |
| `io.confluent.kafka.serializers.subject.RecordNameStrategy` | Fully-qualfied record name | `<record-name>-key` | `<record-name>-value` |
| `io.confluent.kafka.serializers.subject.TopicRecordNameStrategy` | Both topic name and fully-qualified record name | `<topic-name>-<record-name>-key` | `<topic-name>-<record-name>-value`

:::note
The default subject naming strategy is `io.confluent.kafka.serializers.subject.TopicNameStrategy`.
:::

## Configure schema ID validation

Set Redpanda cluster properties to enable schema ID validation for keys, values, or both:

- Set [`record_key_schema_id_validation`](../../reference/cluster-properties#record_key_schema_id_validation) to `true` to enable key schema ID validation, then set [`record_key_subject_name_strategy`](../../reference/cluster-properties#record_key_subject_name_strategy) to the desired subject name strategy for keys (default: TopicNameStrategy). 

> - Set [`record_value_schema_id_validation`](../../reference/cluster-properties#record_value_schema_id_validation) to `true` to enable value schema ID validation, then set [`record_value_subject_name_strategy`](../../reference/cluster-properties#record_value_subject_name_strategy) to the desired subject name strategy for values (default: TopicNameStrategy). 

Reviewer 2: I'm going to remove these cluster properties, although the defaults will still exist.

To override the cluster configuration for each topic, set a client's topic properties:
- Set `confluent.key.schema.validation` to `true` to enable key schema ID validation for the topic, then set `confluent.key.subject.name.strategy` to the desired subject name strategy for keys of the topic (default: `io.confluent.kafka.serializers.subject.TopicNameStrategy`). 

> - Set `confluent.value.schema.validation` to `true` to enable value schema ID validation for the topic, then set `confluent.value.subject.name.strategy` to the desired subject name strategy for values of the topic (default: `io.confluent.kafka.serializers.subject.TopicNameStrategy`). 

Reviewer 2: These configuration will be aliased:
`confluent.key.schema.validation` -> `redpanda.key.schema.id.validation`
`confluent.key.subject.name.strategy` -> `redpanda.key.subject.name.strategy`
`confluent.value.schema.validation` -> `redpanda.value.schema.id.validation`
`confluent.value.subject.name.strategy` -> `redpanda.value.subject.name.strategy`

Either and both can be accepted configurations, but if both become set (during topic `create` or `alter`), and they are incompatible, the create or update shall be rejected.

Where the configuration for `subject.name.strategy` is prefixed `redpanda`, the available strategies drop the `io.confluent.kafka.serializers.subject.` prefix:
* `TopicNameStrategy`
* `RecordNameStrategy`
* `TopicRecordNameStrategy`

But are otherwise equivalent to the ones that have the `io.confluent.kafka.serializers.subject`.

:::note
Enabling schema ID validation on a compressed topic requires decompressing each batch before it is written.

> :::

Author: @BenPope - need your help to clarify this note. The required decompression is on the producer client? Should schema ID validation not use producer compression? Or is this talking about decompression on the Redpanda broker?

Reviewer 2: This is on Redpanda.

---
title: Server-side Schema ID Validation
linkRoot: ../../
---

<head>
    <meta name="title" content="Server-side Schema ID Validation | Redpanda Docs"/>
    <meta name="description" content="Learn about server-side schema ID validation for clients using Confluent's serdes format that produce to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy and registered with the Schema Registry."/>
    <link rel="canonical" href="https://docs.redpanda.com/manage/schema-registry/" />
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import EnterpriseLicenseNote from '@site/docs/shared/_enterprise-license.mdx'

Learn about server-side schema ID validation for clients using Confluent's serdes format that produce to Redpanda brokers, and learn how to configure Redpanda to inspect and reject records with schema IDs that aren't valid according to the configured Subject Name Strategy and registered with the Schema Registry.

:::info
<EnterpriseLicenseNote frontmatter={frontMatter}/>
:::

## About schema ID validation

> Records produced to a topic may use a serializer/deserializer (cloud_storage_max_materialized_segments_per_shard) client library, such as Confluent's serdes library, to encode their keys and values according to a schema in [Schema Registry](../schema-registry).

Reviewer 2: `cloud_storage_max_materialized_segments_per_shard`??

Author: fixed (copy-paste error)

> When a client produces a record, the schema ID for the topic is encoded in the record's payload header. A client may misconfigure and produce a record whose schema ID doesn't match any schema in Schema Registry. A produced record for an unregistered schema shouldn't be stored by brokers or fetched by consumers. Yet, it may not be detected or dropped until after it's been fetched and a consumer deserializes its mismatched schema ID.

Reviewer 2: > A client may misconfigure and produce a record whose schema ID doesn't match any schema in Schema Registry.

It's not really about whether the schema ID exists in the Schema Registry. The schema ID must be associated with the Subject in the Schema Registry. That Subject is determined by the Subject Name Strategy.

A client may be misconfigured with either the wrong schema, or wrong subject name strategy, resulting in unexpected data on the topic.

Author: Reworded

To enable schema ID validation, clients must use a supported **subject naming strategy** that maps the topic and schema onto a subject. When a client produces a record, the schema ID is encoded into the header of a record field, where a record field is the key or value of that record. The topic and schema record are combined according to the subject name strategy to form a subject. The schema ID is associated with a subject and version. 

> **Schema ID validation enables brokers (servers) to detect and drop records that were produced with an incorrectly configured subject name strategy, that don't conform to the serdes wire format, or encode an incorrect schema ID**. With schema ID validation, records associated with unregistered schemas are detected and dropped earlier, by a broker rather than a consumer. 

Reviewer: See https://github.com/redpanda-data/documentation/blob/dev/meta_docs/style_guide.mdx#bold-text
If it is really important--maybe use an important admonition?

:::info
Schema ID validation doesn't verify that a record's payload is correctly encoded according to the associated schema. Schema ID validation only checks that the schema ID encoded in the record is registered in the Schema Registry.
:::

## Configure schema ID validation

To use schema ID validation:
- [Enable the feature in Redpanda](#enable-schema-id-validation)
- [Customize the subject name strategy per topic on the client](#set-subject-name-strategy-per-topic)

### Enable schema ID validation

To enable schema ID validation in Redpanda, set it with the `/v1/features/schema_id_validation` Admin API endpoint:

```bash

> curl -sv -X PUT <broker-address>:9644/v1/features/schema_id_validation -d '{"state":"active"}'

Reviewer 2: This is on by default, but it's useful to know which flag this feature is guarded by. @mattschumpert did we want this on by default? It means all the topic properties are listed during a describe.

```

### Set subject name strategy per topic

The subject naming strategies supported by the Redpanda:

| Subject Name Strategy | Subject Name Source | Subject Name Format (Key) | Subject Name Format (Value) |
|- | - | - | - |
| `TopicNameStrategy` | Topic name | `<topic-name>-key` | `<topic-name>-value` |

> | `RecordNameStrategy` | Fully-qualfied record name | `<record-name>-key` | `<record-name>-value` |
> | `TopicRecordNameStrategy` | Both topic name and fully-qualified record name | `<topic-name>-<record-name>-key` | `<topic-name>-<record-name>-value`

Reviewer 2: The key/value suffix is only for `TopicNameStrategy`
```suggestion
| `RecordNameStrategy` | Fully-qualfied record name | `<record-name>` | `<record-name>` |
| `TopicRecordNameStrategy` | Both topic name and fully-qualified record name | `<topic-name>-<record-name>` | `<topic-name>-<record-name>`
```

When [schema ID validation is enabled](#enable-schema-id-validation), Redpanda uses `TopicNameStrategy` by default.

To customize the subject name strategy per topic, set the following client topic properties:

- Set `redpanda.key.schema.id.validation` to `true` to enable key schema ID validation for the topic, then set `redpanda.key.subject.name.strategy` to the desired subject name strategy for keys of the topic (default: `TopicNameStrategy`). 

> - Set `redpanda.value.schema.id.validation` to `true` to enable value schema ID validation for the topic, then set `redpanda.value.subject.name.strategy` to the desired subject name strategy for values of the topic (default: `TopicNameStrategy`). 

Reviewer 2: "then" implies an ordering, I think "and" is better, they can be set at the same time either during topic creation or with alter topic.

Reviewer 2: An example might be useful:
Create:
```
rpk topic create topic_foo \
  --topic-config redpanda.value.schema.id.validation=true \
  --topic-config redpanda.value.subject.name.strategy=RecordNameStrategy \
  --brokers 127.0.0.1:9092
```
Alter:
```
rpk topic alter-config topic_foo \
  --set redpanda.value.schema.id.validation=true \
  --set redpanda.value.subject.name.strategy=RecordNameStrategy \
  --brokers 127.0.0.1:9092
```

Author: Added

:::info

> The `redpanda.*` properties are aliases of corresponding `confluent.*` properties.

Reviewer 2: Is not really an alias any more, they're distinct, but if both are set, they must be equivalent.

Author: Fixed

- [Enable the feature in Redpanda](#enable-schema-id-validation)
- [Customize the subject name strategy per topic on the client](#set-subject-name-strategy-per-topic)

### Enable schema ID validation

To enable schema ID validation in Redpanda, set it with the `/v1/features/schema_id_validation` Admin API endpoint:

```bash
curl -sv -X PUT <broker-address>:9644/v1/features/schema_id_validation -d '{"state":"active"}'
```

### Set subject name strategy per topic

The subject naming strategies supported by the Redpanda:

| Subject Name Strategy | Subject Name Source | Subject Name Format (Key) | Subject Name Format (Value) |
|- | - | - | - |
| `TopicNameStrategy` | Topic name | `<topic-name>-key` | `<topic-name>-value` |
| `RecordNameStrategy` | Fully-qualfied record name | `<record-name>-key` | `<record-name>-value` |
| `TopicRecordNameStrategy` | Both topic name and fully-qualified record name | `<topic-name>-<record-name>-key` | `<topic-name>-<record-name>-value`

When [schema ID validation is enabled](#enable-schema-id-validation), Redpanda uses `TopicNameStrategy` by default.

To customize the subject name strategy per topic, set the following client topic properties:

> - Set `redpanda.key.schema.validation` to `true` to enable key schema ID validation for the topic, then set `redpanda.key.subject.name.strategy` to the desired subject name strategy for keys of the topic (default: `TopicNameStrategy`). 

Reviewer 2: You dropped an `.id` from my previous review comment.

Author: @BenPope after https://github.com/redpanda-data/redpanda/pull/10650 is approved, please review/approve this
